include/El/optimization/bfgs.hpp

```diff
@@ -0,0 +1,226 @@
+                return alpha;
+            }
+            if( fval_dash > 0)
+            {
+                return Zoom(f, gradient, f0, x0, p, alpha, alpha_prev, c1, c2);
+            }
+            alpha_prev = alpha;
+            fvalPrev = fval;
+            alpha = std::min(2*alpha, alphaMax);
+        }
+    return alpha;
+}
+
+namespace detail {
+/***
+ * This class stores the updates to the inverse hessian of F, and provides the ability to apply inverse hessian.
+ */
+template< typename T>
+struct HessianInverseOperator {
+
+    typedef DistMatrix<T> Matrix;
+    typedef std::tuple< T, Matrix, Matrix, Matrix, T> Update;
+    typedef std::vector< Update > Hessian_updates;
+    /**
+     * This method applies y = H_k*x, eg. solves B_ky = x;
+     * (H_k)x = z + alpha*(s'x)s - beta*H_{k-1}y- rho*s
+     * where alpha = (s'y + y'H_{k-1}y)/(s'y)^2
+     * where beta = (s'x)/(s'y)
+     * where rho = (y'z)/(s'y)
+     *
+     * @param x
+     * @return H_k*x
+     */
+    Matrix operator*( const Matrix& x)
+    {
+        //Initially this is just the identity matrix;
+        //H_0 = I, so H_0*x = x;
+        if (_hessian_data.size() == 0) { return x; }
+        DistMatrix<T> z( x);
+
+        //we maintain that z = H_{k-1}*x
+        // H_kx = z + alpha*(s'x)s - (s'x)/(s'y)H_{k-1}y  (y'z)/(s'y)*s]/(s'y)
+        for( auto& rank_two_update_data: _hessian_data)
+        {
+            auto alpha = std::get<0>(rank_two_update_data);
+            const auto& s = std::get<1>(rank_two_update_data);
+            const auto& y = std::get<2>(rank_two_update_data);
+            const auto& Hy = std::get<3>(rank_two_update_data);
+            const auto& sy = std::get<4>(rank_two_update_data);
+            //s'x
+            auto sx = Dot(s,x);
+            //y'z
+            auto yz = Dot(y,z);
+            auto beta = -sx/sy;
+            auto rho = -yz/sy;
+            // (H_k)x = z + alpha*(s'x)s + beta*H_{k-1}y + rho*s
+            //We now view:
+            //  (H_k)x = z + alpha*s + beta*Hy + rho*s
+            //         = z + (alpha+rho)*s - beta*Hy
+            // Rank-1 update: z <- z+(alpha+rho)*s
+            Axpy((alpha*sx+rho), s, z);
+            //Rank-1 update: z <- z + beta*Hy
+            Axpy( beta, Hy, z);
+        }
+        return z;
+    }
+    /**
+     * The forward update is:
+     * B_{k+1} = B_k + yy'/(y's) + B_ksks'B_k/(s'Bs)
+     * So the backwards update is:
+     * H_{k+1} = H_k + alpha*(ss') - [H_kys' + sy'H_k]/(s'y)
+     * where alpha = [(s'y + y'H_ky)/(s'y)^2];
+     * This method advances the operator to represent H_{k+1}
+     * @param s
+     * @param y
+     */
+    void update( DistMatrix<T>& s, DistMatrix<T>& y)
+    {
+        // We store s,y, alpha, H*y, and s'y.
+        auto sy = Dot(s,y);
+        auto Hy = (*this)*y;
+        auto yHy = Dot(Hy,y);
+        T alpha = (sy + yHy)/(sy*sy);
+        _hessian_data.emplace_back( alpha, s, y, Hy, sy);
+    }
+
+private:
+    Hessian_updates _hessian_data;
```

Comment:
It is usually considered a bad idea to prefix an underscore, as variables which start with an underscore and are followed by a capital letter are reserved for internal implementations (and I have been burned by this in the past). For this reason, it is better to postfix the underscore to avoid later problems.